Add synthetic value class companion near its class.

hrhino opened the pull request:
Redux of 9e1de6ee81e9eaf9d8ac59446bc97c79b5ff0cb6.

Make sure that the class in question actually exists in the tree we're about to put the synthetic companion in. Otherwise extmethods might not see the extendable methods until after it's too late to add it to the companion stats.

Fixes scala/bug#10783.

hrhino committed:
Add synthetic value class companion near its class.

Redux of 9e1de6ee81e9eaf9d8ac59446bc97c79b5ff0cb6.

Make sure that the class in question actually exists in the
tree we're about to put the synthetic companion in. Otherwise
extmethods might not see the extendable methods until after
it's too late to add it to the companion stats.

Fixes scala/bug#10783.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3214,6 +3216,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+                  stats.exists {
+                    case t @ ClassDef(_, _, _, _) => t.symbol eq csym
+                    case _ => false
+                  }
```
Would it be too dry to refactor to `=> classInStats(_.symbol == sym.companionClass)` and the other case `classInStats(_.symbol == att.caseClass.symbol)` or what have you?

hrhino replied:
Good call. I'm absorbing so much information at NEScala that I've got little chance of being (metaphorically) DRY.

som-snytt reviewed (approved):
Missed opportunity for single line cases, but oh well. :)

hrhino commented:
The `case class` case pushes the line past 80 chars, and 4-deep nested methods don't excuse that. (Comment lines don't count, I guess).

som-snytt commented:
Adriaan likes long lines. `class Typer` is 135 columns. Also `SilentTypeError`. `adaptToArguments` at 147. I may get one of those wraparound gaming/coding monitors.

hrhino commented:
Evidently so. `TypeTestTreeMaker` is 203, and that's after only a cursory search. The tradeoff is, of course, against the length of the file, like good old `Typers.scala`. No one said typing Scala was easy, though.

Do you mean one of [these](https://gizmodo.com/5926043/175-degree-display-fills-your-entire-vision-with-gaming-goodness)? My laptop fits two-column github view mode with 85 columns or so, depending whether my browser's in fullscreen mode or no. I wonder if my company would let me comp one of those. I wonder if it's because Lightbend got him one.

hrhino commented:
/sync

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3205,15 +3205,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+          // scala/bug#10783 ditto for synthetic companions of derived value classes.
           def shouldAdd(sym: Symbol): Boolean = {
+            def classDefinedHere(s: Symbol): Boolean = stats exists {
+              case t: ClassDef => t.symbol eq s
```
Not sure if it matters, but this will miss `ClassDefs`-s that are wrapped in `DocDef`-s when running under scalacheck mode. The painful ease with which one can make this mistake is lamented in https://github.com/scala/scala/blob/9acab45aeeadef2f63da69faf81465cc15599789/src/compiler/scala/tools/nsc/ast/TreeInfo.scala#L76-L107

hrhino replied:
Do you mean scaladoc mode? I did think about that but scaladoc working differently here _probably_ won't make a difference. I could make a `DocDef(_, t: ClassDef)` case.

som-snytt replied:
A scaladoc test should already demonstrate the existing issue.

retronym commented on test/files/run/t10783.scala:
```diff
@@ -0,0 +1,24 @@
+  class Value(val value: Int) extends AnyVal {
+    def foo: Int = value + 1
+  }
+}
```
Test variations with an explicitly written companion, defined before and after this class, would be welcome.

hrhino replied:
Sounds cautious. Added.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3205,15 +3205,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-                  }
+                  classDefinedHere(att.caseClass.symbol)
+                case _ if sym.companionClass.isDerivedValueClass =>
+                  classDefinedHere(sym.companionClass)
```
It seems wrong to need a special case for value classes here, rather than being able to generically detect that `ensureCompanionObject` was called and no explicit companion was in the vicinity.

e.g. default arguments for constructors give rise to synthetic companions to host the default getters. And compiler plugins might use `ensureCompanionObject` for custom, but similar, use cases.

hrhino replied:
I fiddled with the default args case and couldn't get weirdness to occur, although it does go into the wrong tree for

```scala
package fizz {}
package fizz { class Foo(i: Int = 1) }
```

So that's something that should probably be avoided.

I spent a plane ride convincing myself that `att.caseClass.symbol eq companionSymbol(sym, context)` and I think I've succeeded. So we can probably safely have

```scala
def shouldAddAsModule = classDefinedHere(companionSymbolOf(sym, context))
```

(As warned about, `sym.companionClass` fails for (e.g.) `pos/t911.scala`).

som-snytt replied:
The default getters case was mentioned by contributor. Isn't the general solution to use attachments for all trees with associated synthetics? Currently we have synthetic syms and ask if they belong in the current stats, instead of looking at the stats for things with synthetics. In my second contribution, I punned on ad-hoc as add-hack, that was a good one. https://github.com/scala/scala/commit/ed723738e643035609efda49e6be689e4b0b0480

som-snytt replied:
(Obviously, I wrote my reply before I saw hrhino's.)

retronym reviewed (changes requested):
Marking as "Request changes" to consider the `DocDef` and general use of `ensureCompanionObject` questions.

retronym commented:
We could rule that problem out by requiring the companions, even package owned ones, should be co-defined in a scope. Does that sound too onerous? We could phase it in with a deprecation.

A double traversal in extmethods wouldn't be too horrible, as the first one would only need to be shallow (because value classes may only be top level).

Adding cross references to other trees via symbol or tree attachments is a tool of last resort, it is all too easy some intervening transform to update the tree leaving a stale one in the attachment.

hrhino commented:
That doesn't seem very onerous at all, but technically the spec only disallows them in different scopes, so it may require a tweak to that language.

Any feature which requires moving trees into companions post-typer would either break here or provide guidance on how to continue. (Constructor default getters are done in namer so they appear to be exempt.) I'm trying to think of another such beast.

I'll prototype the double traversal and we can see if it hurts performance. This is a pretty minor edge case, either way.

som-snytt commented:
It seems to me that if there is an attachment, it's easier to just use it, rather than doing look-ups. Case classes are common; I don't know when `companionSymbolOf` gets expensive.

som-snytt commented:
@hrhino The issue is still labelled "has PR", hint hint.

hrhino commented:
Winks, hints, judges, &c. acknowledged and accepted.

 I forgot how many alternatives came up during review. I'll consider them all with a fresh mind. Maybe I'll get a friend to write them down anonymously, for double-blind self-review.

hrhino commented:
While I re-familiarize myself with what I did here, I decided to post on the contributors' Discourse regarding Jason's suggestion above: https://contributors.scala-lang.org/t/-/2122.

By doing this, I've probably ensured that it will never be implemented ("make it out of committee, as we say here"), but it was worth a go.

hrhino commented:
@jvican thank you! FWIW, this specific bug can be fixed in ways other than that, and I personally think it's cleaner to go with an approach more like the one that's here right now.

Thank you for taking such a quick look at this, though! 🙏

jvican commented:
So, I guess no SIP is needed after all! Ping me if you have more mini SIPs you wanna discuss in the future, I'm sure you will :smile:

hrhino commented:
I don't see why not. There doesn't seem to be any un-addressed commentary.

The change to enforce that companions are co-defined can come later. (Are you asking so this can make the last forward merge before M5?)

Also, shh, you're going to reveal @som-snytt's secret identity!

dwijnand commented:
> (Are you asking so this can make the last forward merge before M5?)

Nope, just giving a pass at the 2.12.7 PR queue too, and confused by the state of this.

Ok, then thanks for the PR :)

hrhino commented:
Thanks for the merge! ❤️
